server/weaverbird/backends/pandas_executor/steps/waterfall.py

```diff
@@ -0,0 +1,166 @@
+    return merged_df
+
+
+# the waterfall has a very specific structure.
+# this methods create the top / bottom rows.
+# these contains the sum of values for the whole milestone, regardless of label.
+def compute_agg_milestone(
+    step: WaterfallStep, df: DataFrame, start_df: DataFrame, milestone
+) -> pd.DataFrame:
+    if len(step.groupby) > 0:
+        groups = df[step.groupby].drop_duplicates()
+        group_by = step.groupby
+    else:
+        groups = pd.DataFrame({'_VQB_GROUP': [0]})  # pseudo group
+        start_df['_VQB_GROUP'] = 0
+        groups['_VQB_GROUP'] = 0
+        group_by = ['_VQB_GROUP']
+
+    groups = groups.assign(**{step.labelsColumn: milestone})
+
+    agg = start_df.groupby(group_by).agg({f'{step.valueColumn}': 'sum'})
+    agg = (
+        groups.merge(agg, on=group_by)
+        .sort_values(
+            by=step.valueColumn if step.sortBy == 'value' else step.labelsColumn,
+            ascending=step.order == 'asc',
+        )
+        .rename(columns={step.labelsColumn: LABEL_WATERFALL_COLUMN})
+    )
+
+    if step.parentsColumn is not None:
+        agg[GROUP_WATERFALL_COLUMN] = agg[LABEL_WATERFALL_COLUMN].astype(str)
+    agg[TYPE_WATERFALL_COLUMN] = None
+    agg[LABEL_WATERFALL_COLUMN] = agg[LABEL_WATERFALL_COLUMN].astype(str)
+    if len(step.groupby) == 0:
+        del start_df['_VQB_GROUP']
+        del agg['_VQB_GROUP']
+    return agg
+
+
+# this method shapes the merged DF so it matches the waterfall format
+def compute_mid(step: WaterfallStep, merged_df: DataFrame, df: DataFrame) -> DataFrame:
+    result_df = DataFrame(
+        {
+            LABEL_WATERFALL_COLUMN: merged_df.sort_values(
+                by=get_sort_column(step), ascending=step.order == 'asc'
+            )[step.labelsColumn].astype(str)
+        }
+    )
+    result_df[step.groupby] = merged_df.sort_values(
+        by=get_sort_column(step), ascending=step.order == 'asc'
+    )[step.groupby]
+
+    if step.parentsColumn is None:
+        result_df[TYPE_WATERFALL_COLUMN] = 'Parent'
+    else:
+        result_df[GROUP_WATERFALL_COLUMN] = merged_df.sort_values(
+            by=get_sort_column(step), ascending=step.order == 'asc'
+        )[step.parentsColumn]
+        result_df[TYPE_WATERFALL_COLUMN] = numpy.where(
+            result_df['LABEL_waterfall'] == (result_df[GROUP_WATERFALL_COLUMN]),
+            ['parent'],
+            ['child'],
+        )
+
+    result_df[step.valueColumn] = merged_df.sort_values(
+        by=get_sort_column(step), ascending=step.order == 'asc'
+    )[_RESULT_COLUMN]
+    return result_df
+
+
+def get_sort_column(step: WaterfallStep):
+    if step.sortBy == 'value':
+        return _RESULT_COLUMN
+    else:
+        return step.labelsColumn
+
+
+def get_join_key(step: WaterfallStep):
+    if step.parentsColumn is None:
+        return [step.labelsColumn] + step.groupby
+    else:
+        return [step.labelsColumn, step.parentsColumn] + step.groupby
```

Comment:
You've made it to the end of steps :tada: :tada: :tada: :tada: :tada:  big thanks and congrats!